src/tla/__rewire_sequences_ext_in_apalache.tla

```diff
@@ -0,0 +1,470 @@
+ *)
+Prefixes(s) ==
+  { SubSeq(s, 1, __l): __l \in { 0 } \union DOMAIN s }
+
+(**
+ * The set of all sequences that are prefixes of the set of sequences S.
+ *
+ * @type: Set(Seq(a)) => Set(Seq(a));
+ *)
+CommonPrefixes(S) ==
+  \* TODO: use FoldSet?
+  LET __P == UNION { Prefixes(seq) : seq \in S }
+  IN { __prefix \in __P: \A __t \in S: IsPrefix(__prefix, __t) }
+
+(**
+ * The longest common prefix of the sequences in the set S.
+ *
+ * @type: Set(Seq(a)) => Seq(a);
+ *)
+LongestCommonPrefix(S) ==
+  CHOOSE __longest \in CommonPrefixes(S):
+      \* there can only be one LCP => CHOOSE
+      \A __other \in CommonPrefixes(S):
+          Len(__other) <= Len(__longest)
+
+-----------------------------------------------------------------------------
+
+(**
+ * Range(a % b) = 0..b-1, but DOMAIN seq = 1..Len(seq).
+ * So to do modular arithmetic on sequences we need to map 0 to b.
+ *
+ * @type: (Int, Int) => Int;
+ *)
+SeqMod(a, b) == 
+  IF a % b = 0 THEN b ELSE a % b
+
+(**
+ * An alias of FoldFunction that op on all elements of seq an arbitrary
+ * order. The resulting function is:
+ *
+ *    op(f[i],op(f[j], ..., op(f[k],base) ...))
+ *
+ * op must be associative and commutative, because we can not assume a
+ * particular ordering of i, j, and k
+ *
+ * Example:
+ *
+ *  FoldSeq(LAMBDA x,y: {x} \cup y, {}, <<1,2,1>>) = Range(<<1,2,1>>)
+ *
+ * @type: ((a, b) => b, b, Seq(a)) => b;
+ *)
+FoldSeq(op(_, _), base, seq) == 
+  \* __ApalacheFoldSeq is accumulating the result in the left argument,
+  \* whereas FoldSeq is accumulating the result in the right argument.
+  LET __map(__y, __x) == op(__x, __y) IN
+  __ApalacheFoldSeq(__map, base, seq)
+
+(**
+ * FoldLeft folds op on all elements of seq from left to right, starting
+ * with the first element and base. The resulting function is:
+ *    op(op(...op(base,f[0]), ...f[n-1]), f[n])
+ *
+ * Example:
+ *    LET cons(x,y) == <<x,y>>
+ *    IN FoldLeft(cons, 0, <<3, 1, 2>>) = << << <<0, 3>>, 1>>, 2>> 
+ *
+ * @type: ((b, a) => b, a, Seq(a)) => b;
+ *)
+FoldLeft(op(_, _), base, seq) == 
+  LET __map(__x, __y) == op(__x, __y) IN
+  __ApalacheFoldSeq(__map, base, seq)
+
+(**
+ * FoldRight folds op on all elements of seq from right to left, starting
+ * with the last element and base. The resulting function is:
+ *    op(f[0],op(f[1], ..., op(f[n],base) ...))
+ *
+ *
+ * Example:
+ *    LET cons(x,y) == <<x,y>>
+ *    IN FoldRight(cons, <<3, 1, 2>>, 0 ) = << 3, << 1, <<2, 0>> >> >> 
+ *
+ * @type: ((a, b) => b, Seq(a), b) => b;
+ *)
+FoldRight(op(_, _), seq, base) == 
+  LET __map(__y, __x) == op(__x, __y) IN
+  __ApalacheFoldSeq(__map, base, Reverse(seq))
+
+-----------------------------------------------------------------------------
+
+(**
+ * A sequence of all elements from all sequences in the sequence seqs.
+ *
+ * Examples:
+ *
+ *  FlattenSeq(<< <<1, 2>>, <<1>> >>) = << 1, 2, 1 >>
+ *  FlattenSeq(<< <<"a">>, <<"b">> >>) = <<"a", "b">>
+ *  FlattenSeq(<< "a", "b" >>) = "ab"
```

Comment:
Will this work in Apalache, i.e., is `Str` a `Seq`?

Reply:
No, but AFAIK Igor keeps the original comments

Reply:
nope. Strings in Apalache are indivisible. I have to fix this comment.